Add tests for the propagation of vulnerabilities within git repositories

CharlyReux opened the pull request:
In relation to some research we are making about the security in the Open Source ecosystem, we investigated how vulnerabilities are propagated in repositories. While reviewing the analysis output of the propagation of vulnerability, we noticed some unexpected behaviors, as per  @RomainLefeuvre 's [issues](https://github.com/google/osv.dev/issues/1910).

We made a documentation of the behavior for the test of repositories that you can find below. 
We followed what we interpreted from the [OSV schema documentation](https://ossf.github.io/osv-schema/), and we would like to know whether this documentation suits it, and more precisely whether the test cases make sense to you. 

![test_cases_osvdev](https://github.com/google/osv.dev/assets/95578339/998a623b-f3cf-4c5c-986b-a2b4fb7b5643)



This PR contains the implementation of a new test class impact_git_test and a tool to facilitate the creation of git repositories programmatically, the repositories created are dummy repositories, containing empty commits. 

The tests implemented test the repoanalyzer class, without using the cherry-pick detection.

andrewpollock commented:
/gcbrun

andrewpollock commented on osv/test_tools/test_repository.py:
```diff
@@ -0,0 +1,158 @@
+  """ Utilitary class to create a test repository for the git tests
+  """
+
+  class VulnerabilityType(Enum):
```
For future-proofing and maintenance, could you try and use `osv.vulnerability_pb2` and use `vulnerability_pb2.Event.DESCRIPTOR.fields_by_name` to introspect into the OSV record definition code, rather than having this standalone definition that could go out of sync with what's defined in https://github.com/google/osv.dev/blob/ad0eed1cdb155bedbbe195f049b32daf459deca5/osv/vulnerability.proto#L67-L77

See also https://googleapis.dev/python/protobuf/latest/google/protobuf/descriptor.html

andrewpollock commented on osv/test_tools/test_repository.py:
```diff
@@ -0,0 +1,158 @@
+
+
+class TestRepository:
+  """ Utilitary class to create a test repository for the git tests
```
Fix spelling mistake

```suggestion
  """ Utility class to create a test repository for the git tests
```

andrewpollock commented on osv/test_tools/test_repository.py:
```diff
@@ -0,0 +1,158 @@
+      shutil.rmtree(f"osv/testdata/test_repositories/{name}")
+    self.repo = pygit2.init_repository(
+        f"osv/testdata/test_repositories/{name}", bare=False)
+    #empty initial commit usefull for the creation of the repository
```
Fix spelling mistake

```suggestion
    #empty initial commit useful for the creation of the repository
```

andrewpollock commented on osv/test_tools/test_repository.py:
```diff
@@ -0,0 +1,158 @@
+"""test_repository"""
```
Please expand this with more details, see https://google.github.io/styleguide/pyguide.html#381-docstrings

andrewpollock reviewed (commented):
Thank you for this contribution, apologies for the slow initial review. My review is still ongoing, here's some initial feedback.

I'd like @oliverchang to also review this, as he's much more intimately familiar with the code being tested here, but I will give this my best shot along.

andrewpollock commented:
I note that this introduces breakage, so that will need to be resolved (perhaps the point of this PR was drive that discussion) before this can be merged successfully...

CharlyReux commented:
Thank you very much for the review, and I apologize for the delay in responding. Since the initial PR commit, we have undergone some refactoring, mainly to make the code easier to understand, implement your proposed changes, and support the cherry-pick parameter to true when using `RepoAnalyzer`. Additionally, we reviewed the test cases we had proposed and noticed some inconsistencies, so we made slight modifications to them. 


The aim of this pr is indeed to start the discussion about the possible discrepancy in the osv.dev implementation with regard to the "Open Source Vulnerability format"  specification. For a research project, we implemented a test suite of the vulnerable commit identification algorithm to test our implementation. We thought that osv.dev would benefit from the same tests.

We identified discrepancies while using OSV.dev for a research project and created associated issue : 

- #1910
- #1938

These test cases were implemented based on our interpretation of the specification. The specification seems to be ambiguous for certain situations, such as merge propagation, multiple ranges ...  


We hope that the tests we implemented, might help to make the specification less ambiguous. 

![osv_test_cases drawio](https://github.com/google/osv.dev/assets/95578339/394aadbe-55a4-4de4-b116-965c66b483b3)

RomainLefeuvre commented:
The objective of this PR was indeed to drive the discussion about how to test the affected git commit detection and avoid ambiguity on the specification.  Do not hesitate if you have any feedback.

andrewpollock commented:
We want to get around to spending some time on this, it's just been a case of insufficient time and higher priorities :-(

RomainLefeuvre commented:
I understand that GIT vulnerability algorithm is not a top priority, as most of OSVs rely on Semantic Versioning or ecosystem specific version format. 

However, I'm convinced that pushing for a broader adoption of GIT RANGES should enable many analyses that are not possible with Semantic Versionning ranges but also increase the quality of vulnerable commit/release/version identification.

 For a research project, we developed an algorithm that label the Software Heritage Graph using as input OSV having a git range.
 If you are interested, we can discuss the different test cases and formalize on the _Open Source Vulnerability format_  specification an evaluation algorithm adapted to GIT ranges.

oliverchang commented:
Apologies for letting this languish. We'll take a look some time in the future when we get the cycles to improve our git handling!
